### ISSUE TITLE
[GSoC 2025] Apply quantization to reduce model footprint

### ISSUE THREAD

**Issue by singjc** [Roestlab/diffusion-deconvolution-dia-msms-data#17]
## Background
Our diffusion model for DIA-MS data deconvolution currently has a large memory footprint, which limits our ability to process larger datasets and deploy the model efficiently. Quantization techniques, which have been successfully applied to vision models, offer a promising approach to reduce model size while maintaining performance.

## Task Objectives

- Implement post-training quantization for our existing diffusion model
- Evaluate multiple quantization strategies (e.g., INT8, FP16) for their impact on:
   - Memory usage
   - Inference speed
   - Model accuracy/performance
- Implement the most effective quantization strategy in our training and inference pipelines

## Technical Details

- The model is implemented in PyTorch with a U-Net-based architecture
- We need to balance memory efficiency with the precision required for accurate MS signal deconvolution
- Begin with static quantization of pre-trained models, then explore quantization-aware training if results are promising

## Deliverables

- Implementation of quantization methods in the model code
- Comprehensive benchmarks comparing original vs. quantized models including:
   - Memory usage measurements
   - Inference time comparisons
   - Quality metrics on test datasets
- Documentation of the quantization process and best practices
- Pull request with the optimized model implementation

## Resources

- [PyTorch Quantization Documentation](https://pytorch.org/docs/stable/quantization.html)
- [Model architecture code](https://github.com/Roestlab/diffusion-deconvolution-dia-msms-data/tree/main/dquartic/model)
- [Memory-efficient Diffusion Transformers with Quanto and Diffusers ](https://huggingface.co/blog/quanto-diffusers)
- [Q-Diffusion: Quantizing Diffusion Models (2023)](https://arxiv.org/abs/2302.04304)
- [QuEST: Low-bit Diffusion Model Quantization via Efficient Selective Finetuning (2024)](https://arxiv.org/abs/2402.03666)

## Difficulty
Intermediate - Requires understanding of both the model architecture and quantization techniques.

**Comment by sharon-kathambi** [comment]
Hello @singjc I would like to take on this issue and work on it

**Comment by singjc** [comment]
> Hello [@singjc](https://github.com/singjc) I would like to take on this issue and work on it

Hi @sharon-kathambi , thank you for the interest in the project! Of course, we'd be happy to have you contribute. I just wanted to mention though, there is no rush to start coding. The current phase of GSoC 2025 is still the introductory community building phase, so it will be good to get familiar with the code base and literature.

See [GSoC contribution](https://openms.de/news/gsoc2025/#gsoc-contributors) and [GSoC 2025 timeline](https://developers.google.com/open-source/gsoc/timeline)

**Comment by CoderOMaster** [comment]
@singjc hey can assign me the issue, I wanna see if I can make it work by quantization?

**Comment by CoderOMaster** [comment]
@singjc to do PTQ i might need model checkpoints/weight that u might have trained before right? can you please provide that

**Comment by CoderOMaster** [comment]
![Image](https://github.com/user-attachments/assets/2dcd1fec-5f20-4449-9d38-e5220481e245)
@singjc is this snippet fine to load x rows at a time from the db? i tried with 10 but it didnt work ( i am using borrowed uni workstation with 32gb ram and 20gb vram of rtx 4000). so i repalced it with 1 and its taking quite some time. any suggestions?

**Comment by CoderOMaster** [comment]
What could be the reason for having issues with data loading even on decent configuration? Also at the moment ptq quantization is possible I am implementing that and then I will move forward to other techniques like quantization aware training

**Comment by singjc** [comment]
> ![Image](https://github.com/user-attachments/assets/2dcd1fec-5f20-4449-9d38-e5220481e245) [@singjc](https://github.com/singjc) is this snippet fine to load x rows at a time from the db? i tried with 10 but it didnt work ( i am using borrowed uni workstation with 32gb ram and 20gb vram of rtx 4000). so i repalced it with 1 and its taking quite some time. any suggestions?

Strange, I just ran some tests and don't seem to have any issues loading the data using the query limiting (the estimated memory usage is probably not 100% accurate). My machine has 64Gb of ram, but it definitely should need more than 8Gb, especially if you're limiting the number of rows to return by 10. There must be some other issue, maybe IO or something. Is the data physically on the workstation? Or are you reading the data over a network?

```
Limit: 5
Shape of data: (5, 6)
Execution Time: 0.0181 seconds
Memory Used (RSS): 16.88 MB
------------------------------
Limit: 50
Shape of data: (50, 6)
Execution Time: 0.0028 seconds
Memory Used (RSS): 0.00 MB
------------------------------
Limit: 500
Shape of data: (500, 6)
Execution Time: 0.0027 seconds
Memory Used (RSS): 0.16 MB
------------------------------
Limit: 5000
Shape of data: (5000, 6)
Execution Time: 0.0075 seconds
Memory Used (RSS): 0.47 MB
------------------------------
Limit: 50000
Shape of data: (34932, 6)
Execution Time: 0.0258 seconds
Memory Used (RSS): 4.69 MB
------------------------------
Limit: 500000
Shape of data: (34932, 6)
Execution Time: 0.0103 seconds
Memory Used (RSS): 1.56 MB
------------------------------
```

I used this script to test: [dquartic/utils/test_duckdb_parquet_fetching.py](https://github.com/Roestlab/diffusion-deconvolution-dia-msms-data/blob/main/dquartic/utils/test_duckdb_parquet_fetching.py)

**Comment by CoderOMaster** [comment]
@singjc these test cases worked for me too and yes i am loading the data locally. i think there is some issue with the data_loader.py file?

`Limit: 5
Shape of data: (5, 6)
Execution Time: 0.0161 seconds
Memory Used (RSS): 11.01 MB
------------------------------
Limit: 50
Shape of data: (50, 6)
Execution Time: 0.0000 seconds
Memory Used (RSS): 0.13 MB
------------------------------
Limit: 500
Shape of data: (500, 6)
Execution Time: 0.0158 seconds
Memory Used (RSS): 0.05 MB
------------------------------
Limit: 5000
Shape of data: (4352, 6)
Execution Time: 0.0035 seconds
Memory Used (RSS): 0.30 MB
------------------------------
Limit: 50000
Shape of data: (4352, 6)
Execution Time: 0.0015 seconds
Memory Used (RSS): 0.52 MB
------------------------------
Limit: 500000
Shape of data: (4352, 6)
Execution Time: 0.0037 seconds
Memory Used (RSS): 0.02 MB
Memory Used (RSS): 0.52 MB
------------------------------
Limit: 500000
Shape of data: (4352, 6)
Execution Time: 0.0037 seconds
Memory Used (RSS): 0.02 MB
Limit: 500000
Shape of data: (4352, 6)
Execution Time: 0.0037 seconds
Memory Used (RSS): 0.02 MB
Execution Time: 0.0037 seconds
Memory Used (RSS): 0.02 MB
Memory Used (RSS): 0.02 MB
`

**Comment by Omar-Ayman10** [comment]
Hello @singjc, can you assign me this issue? I started checking the recommended literature [here](https://github.com/Roestlab/diffusion-deconvolution-dia-msms-data/wiki/Google-Summer-of-Code-2025) and this issue seems like a good start for me. I have a strong background in Deep Learning and Computer Vision, with hands-on experience developing and fine-tuning cutting-edge models. One of my projects involved building a brain tumor segmentation system using a Swin-Transformer-based U-Net architecture, which has some similarities with this project in terms of its architecture. Looking forward to hearing from you!
